car_test.go

```diff
@@ -234,3 +235,88 @@ func TestEOFHandling(t *testing.T) {
 		}
 	})
 }
+
+func TestBadHeaders(t *testing.T) {
+	testCases := []struct {
+		name   string
+		hex    string
+		errStr string // either the whole error string
+		errPfx string // or just the prefix
+	}{
+		{
+			"{version:2}",
+			"0aa16776657273696f6e02",
+			"invalid car version: 2",
+			"",
+		},
+		{
+			// an unfortunate error because we don't use a pointer
+			"{roots:[baeaaaa3bmjrq]}",
+			"13a165726f6f747381d82a480001000003616263",
+			"invalid car version: 0",
+			"",
+		}, {
+			"{version:\"1\",roots:[baeaaaa3bmjrq]}",
+			"1da265726f6f747381d82a4800010000036162636776657273696f6e6131",
+			"", "invalid header: ",
+		}, {
+			"{version:1}",
+			"0aa16776657273696f6e01",
+			"empty car, no roots",
+			"",
+		}, {
+			"{version:1,roots:{cid:baeaaaa3bmjrq}}",
+			"20a265726f6f7473a163636964d82a4800010000036162636776657273696f6e01",
+			"",
+			"invalid header: ",
+		}, {
+			"{version:1,roots:[baeaaaa3bmjrq],blip:true}",
+			"22a364626c6970f565726f6f747381d82a4800010000036162636776657273696f6e01",
+			"",
+			"invalid header: ",
+		}, {
+			"[1,[]]",
+			"03820180",
+			"",
+			"invalid header: ",
+		}, {
+			// this is an unfortunate error, it'd be nice to catch it better but it's
+			// very unlikely we'd ever see this in practice
+			"null",
+			"01f6",
+			"",
+			"invalid car version: 0",
+		},
+	}
+
+	makeCar := func(t *testing.T, byts string) error {
+		fixture, err := hex.DecodeString(byts)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = NewCarReader(bytes.NewReader(fixture))
+		return err
+	}
+
+	t.Run("Sanity check {version:1,roots:[baeaaaa3bmjrq]}", func(t *testing.T) {
+		err := makeCar(t, "1ca265726f6f747381d82a4800010000036162636776657273696f6e01")
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := makeCar(t, tc.hex)
```

Comment:
Should we check if `err` is not `nil`? Otherwise, I _think_ the code could  panic because of `nil` pointer.

Reply:
👍 good call, have added an `err == nil` check because we always expect an error here and if there isn't one then it's an ... error

Reply:
FYI, another way to do this would be `fmt.Sprint(err)` instead of `err.Error`, then that will give you `<nil>` for nil errors